### PR TITLE
feat: streaming updates checkpoints

### PR DESCRIPTION
**Pull request by eecsliu**
## Description
First pass at streaming updates on checkpoints. Missing down migration and tests, but code compiles. 

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/internal/stream/checkpoints.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+	// step 1: calculate all ids matching this subscription
+	q := db.Bun().
+		NewSelect().
+		Table("checkpoints_v2").
+		Column("checkpoints_v2.id").
+		Join("JOIN trial_id_task_id ON trial_id_task_id.task_id = checkpoints_v2.task_id").
+		Join("JOIN trials ON trial_id_task_id.trial_id = trials.id").
+		Join("JOIN experiments e ON trials.experiment_id = e.id").
+		Join("JOIN projects p ON e.project_id = p.id").
+		OrderExpr("trials.id ASC")
```
[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/stream/checkpoints.go:155]: great point - yeah we'll want to sort by checkpoint id

File: master/static/migrations/20231126215150_add-checkpoints-trigger.tx.up.sql [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+DECLARE
+output jsonb = NULL;
+    temp jsonb = NULL;
+    temptrial jsonb = NULL;
+    tempexp jsonb = NULL;
+BEGIN
+    IF before IS NOT NULL THEN
+        temp = before || jsonb_object_agg('workspace_id', beforework);
+        temptrial = temp || jsonb_build_object('trial_id', trialid);
+        tempexp = temp || jsonb_build_object('experiment_id', expid);
```
Comment (corban-beaird) [inline master/static/migrations/20231126215150_add-checkpoints-trigger.tx.up.sql:35]: nit: it doesn't look like `temptrial` is actually doing anything did you intend for it to be:
```suggestion
        tempexp = temptrial || jsonb_build_object('experiment_id', expid);
```

File: master/internal/stream/test_util.go [hunk 3]
```diff
@@ -41,6 +41,15 @@ func testPrepareFunc(i stream.PreparableMessage) interface{} {
 				typedMsg.ExperimentID,
 				typedMsg.WorkspaceID,
 			)
+		case *CheckpointMsg:
+			return fmt.Sprintf(
```
Comment (corban-beaird) [inline master/internal/stream/test_util.go:45]: nit (non-blocking): could you also check against the trial_id?

**Review by corban-beaird — APPROVED** [review]
Looks great!

**Pull request merged**